Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm

```diff
@@ -0,0 +1,118 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionAPIWindowsEvent.h"
+
+#import "CocoaHelpers.h"
+#import "MessageSenderInlines.h"
+#import "WebExtensionContextMessages.h"
+#import "WebPageProxy.h"
+#import "WebProcess.h"
+#import "_WKWebExtensionUtilities.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+static NSString *windowTypesKey = @"windowTypes";
+static NSString *normalKey = @"normal";
+static NSString *popupKey = @"popup";
+
+namespace WebKit {
+
+void WebExtensionAPIWindowsEvent::invokeListenersWithArgument(id argument, WindowType windowType)
+{
+    if (m_listeners.isEmpty())
+        return;
+
+    for (auto& listener : m_listeners) {
+        if (!listener.second.contains(windowType))
+            continue;
+
+        listener.first->call(argument);
+    }
+}
+
+void WebExtensionAPIWindowsEvent::addListener(WebPage* page, RefPtr<WebExtensionCallbackHandler> listener, NSDictionary *filter, NSString **outExceptionString)
+{
+    static NSArray<NSString *> *optionalKeys = @[
+        windowTypesKey,
+    ];
+
+    static NSDictionary<NSString *, id> *types = @{
+        windowTypesKey: @[ NSString.class ],
+    };
+
+    if (![_WKWebExtensionUtilities validateContentsOfDictionary:filter requiredKeys:nil optionalKeys:optionalKeys keyToExpectedValueType:types outExceptionString:outExceptionString])
+        return;
+
+    OptionSet<WindowType> windowTypeFilter;
+
+    if (NSArray<NSString *> *windowTypes = objectForKey<NSArray>(filter, windowTypesKey, false, NSString.class)) {
+        if ([windowTypes containsObject:normalKey])
+            windowTypeFilter.add(WindowType::Normal);
+        if ([windowTypes containsObject:popupKey])
+            windowTypeFilter.add(WindowType::Popup);
+
+        if (!windowTypeFilter) {
+            *outExceptionString = @"The 'windowTypes' array must contain 'normal' and/or 'popup'.";
```

Comment:
If a developer specified both `normal` and `popup` - would that be the same as not having a filter at all?

Reply:
Yeah, they both end up as `WindowTypes::All` in that case.